### PR TITLE
Store template paths as global variables

### PR DESCRIPTION
**Pull request by joemcgill**
This change updates `locate_template()` so that the stylesheet and template paths get stored to a global variable the first time this function is called and resets these globals any time the theme is switched. This memoizes these values similarly to how the previous `STYLESHEETPATH` and `TEMPLATEPATH` constants were stored prior to https://core.trac.wordpress.org/changeset/56635, but allows these values to be reset when the theme is switched. However, this avoids changing the template path whenever running in a `switch_to_blog` context.

Trac ticket: https://core.trac.wordpress.org/ticket/60290

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/theme.php
Comment (joemcgill) [inline src/wp-includes/theme.php]:
```suggestion
	 * Reset globals to force refresh the next time these directories are accessed 
	 * via `locate_template()` as long as we're not in a switched context.
```

File: src/wp-includes/template.php
Comment (felixarntz) [inline src/wp-includes/template.php]: This memoization makes me worried. We already had memoization for this internally in `get_stylesheet_directory()` and `get_template_directory()`, which got reverted in https://github.com/WordPress/wordpress-develop/commit/b6bf3553d975da2e017a00443b7639aacf0a8586 due to a stream of weird edge-cases.

While not exactly the same, this code effectively reinstantes the memoization for when calling `locate_template()`. Maybe this is fine, but I'm a bit wary about it given the previous problems of the very similar idea.

Given that the ticket is about fixing the problem of using the initial site's directories even when switched on a multisite, it would be safer to avoid the memoization IMO, which partly is a separate performance consideration. We should only access the previously stored values from the initial site when being in a switched state.

The other problem even when doing that would be: What if the first time `locate_template()` is called happens while switched to another site? Then this would store the wrong data. I think to make this reliable we should store the output of these functions in the globals at a point that is determinate in the WordPress bootstrap process, some time earlier when the current theme is loaded and we can be sure we're in the initial site's state (i.e. not switched to another site).

Assuming the initial site's directories get stored in `$wp_stylesheet_path` and `$wp_template_path` globals, then this code here could be something like:
```suggestion
	global $wp_stylesheet_path, $wp_template_path;

	if ( is_multisite() && ms_is_switched() ) {
		$stylesheet_path = $wp_stylesheet_path;
		$template_path   = $wp_template_path;
	} else {
		$stylesheet_path = get_stylesheet_directory();
		$template_path   = get_template_directory();
	}
```

This way there's no memoization, except when in a switched state, in which case there never was.

File: src/wp-includes/theme.php
Comment (felixarntz) [inline src/wp-includes/theme.php]: Better to use the dedicated function:
```suggestion
	if ( ! is_multisite() || ! ms_is_switched() ) {
```

Though per my above comment, we may want to not reset the globals to `null` here, but actually refresh their value right away, since we don't know from which context `locate_template()` will be called next.

**Review by felixarntz — COMMENTED** [review]
@joemcgill Thanks for opening the PR. I have a few points of feedback.

Most importantly, I think we need to ensure the directories stored in the global are always those of the initial site, which right now may not be the case.

File: src/wp-includes/theme.php
Comment (joemcgill) [reply src/wp-includes/theme.php]: Implemented in [8341dbc](https://github.com/WordPress/wordpress-develop/pull/5926/commits/8341dbc4f0a021ce6170c8f91ee1ec796725cf69)

**Comment by joemcgill** [comment]
Thanks for the quick feedback, @felixarntz. After giving it more thought, I had also come to the same conclusion that moving the memoization of the template paths to `load_template()` directly is not ideal, since we cannot be sure when in the request lifecycle that function is being called.

Originally, the template constants were set using the `wp_templating_constants()` function in `wp-settings.php` ([ref](https://github.com/WordPress/wordpress-develop/blob/513992357989158359c4075894bfcb5e5073636b/src/wp-settings.php#L593-L594)), right after the `setup_theme` action is fired as a first step to loading the active theme. Once those constants were set, several templating functions, including `load_template()` relied on those stored values for the rest of the lifecycle.

After [[56635]](https://core.trac.wordpress.org/changeset/56635), those values were instead stored as static variables inside of `get_stylesheet_directory()` and `get_template_directory()`, which were initially called before the theme's `functions.php` file is loaded and any filters provided by the theme were added, which is what caused the initial bugs. By removing that memoization from those functions, we've introduced a new set of issues for the template functions that assumed that relied on the previous constants, as those were immutable (unlike the values returned from the related functions).

As a compromise, I've updated the approach here so that a new function is added, `wp_set_template_globals()` which sets these global values at the same place where the previous constants were being set. This function is also now used in `locate_template()` and `switch_theme()` to ensure globals are properly reset to the currently active theme when needed. We may also need to update other places where the old constants were replaced with the helper functions in the original commit to ensure backwards compatibility.

File: src/wp-includes/template.php
Comment (mukeshpanchal27) [inline src/wp-includes/template.php:684]:
```suggestion
 *
 * @since 6.5.0
 *
```

Missing `@since` annotation.

File: tests/phpunit/tests/theme.php
Comment (mukeshpanchal27) [inline tests/phpunit/tests/theme.php]:
```suggestion
	 * @ticket 60290
	 *
	 * @group ms-required
```

Nit.

File: tests/phpunit/tests/theme.php
Comment (mukeshpanchal27) [inline tests/phpunit/tests/theme.php]:
```suggestion
		// Cleanup.
```

Missing end period.

**Review by mukeshpanchal27 — COMMENTED** [review]
Thanks @joemcgill for the PR. Left some initial feedbacks.

File: src/wp-includes/template.php
Comment (mukeshpanchal27) [inline src/wp-includes/template.php]: Smart 😉 Instead of `is_child_theme()` use local variables.

**Review by mukeshpanchal27 — APPROVED** [review]
Great work, @joemcgill. It looks good from my perspective.

**Comment by zobinadevain** [comment]
<a href='https://capmistoi-6241.weebly.com'>.</a>
<a href='https://capmistoi-6242.weebly.com'>.</a>
<a href='https://capmistoi-6243.weebly.com'>.</a>
<a href='https://capmistoi-6244.weebly.com'>.</a>
<a href='https://capmistoi-6245.weebly.com'>.</a>
<a href='https://capmistoi-6246.weebly.com'>.</a>
<a href='https://capmistoi-6247.weebly.com'>.</a>
<a href='https://capmistoi-6248.weebly.com'>.</a>
<a href='https://capmistoi-6249.weebly.com'>.</a>
<a href='https://capmistoi-6250.weebly.com'>.</a>
<a href='https://capmistoi-6251.weebly.com'>.</a>
<a href='https://capmistoi-6252.weebly.com'>.</a>
<a href='https://capmistoi-6253.weebly.com'>.</a>
<a href='https://capmistoi-6254.weebly.com'>.</a>
<a href='https://capmistoi-6255.weebly.com'>.</a>
<a href='https://capmistoi-6256.weebly.com'>.</a>
<a href='https://capmistoi-6257.weebly.com'>.</a>
<a href='https://capmistoi-6258.weebly.com'>.</a>
<a href='https://capmistoi-6259.weebly.com'>.</a>
<a href='https://capmistoi-6260.weebly.com'>.</a>
<a href='https://capmistoi-6261.weebly.com'>.</a>
<a href='https://capmistoi-6262.weebly.com'>.</a>
<a href='https://capmistoi-6263.weebly.com'>.</a>
<a href='https://capmistoi-6264.weebly.com'>.</a>
<a href='https://capmistoi-6265.weebly.com'>.</a>
<a href='https://capmistoi-6266.weebly.com'>.</a>
<a href='https://capmistoi-6267.weebly.com'>.</a>
<a href='https://capmistoi-6268.weebly.com'>.</a>
<a href='https://capmistoi-6269.weebly.com'>.</a>
<a href='https://capmistoi-6270.weebly.com'>.</a>

**Comment by moshiajonta** [comment]
i was thinking same few days ago this issue is going upper level

**Comment by moshiajonta** [comment]
there are few bottoms outreached by cafnao
https://eventmanagementexpert.com/
https://eventplannertoronto.ca/

File: src/wp-includes/template.php
Comment (joemcgill) [reply src/wp-includes/template.php:684]: Good catch. Holding off on adding this depending on if this gets backported to a 6.4.x minor release.

File: src/wp-includes/template.php
Comment (joemcgill) [reply src/wp-includes/template.php]: Actually, would probably be better to apply these changes to `is_child_theme()` since that function was changed in 6.4 when we deprecated the constants.

Comment (joemcgill) [reply src/wp-includes/template.php]: Applied in [7858833](https://github.com/WordPress/wordpress-develop/pull/5926/commits/7858833b858600e5e6608b9dc67f7a4d3e96782c).

**Comment by joemcgill** [comment]
After reviewing all of the changes that were originally made in [r56635](https://core.trac.wordpress.org/changeset/56635), I believe the only safe way to make this change is to also use of the same globals in all the places where the previous constants were in use. I've applied those changes in the following functions: 

- [is_child_theme](https://github.com/WordPress/wordpress-develop/pull/5926/commits/7858833b858600e5e6608b9dc67f7a4d3e96782c)
- [wp_get_active_and_valid_themes](https://github.com/WordPress/wordpress-develop/pull/5926/commits/fe76be5c1b4a74e75684dd1d89c62534c05ac0fa)
- [resume_theme](https://github.com/WordPress/wordpress-develop/pull/5926/commits/d281cfff6fd52ac6aa056aea7a42120177a18705)
- [comments_template](https://github.com/WordPress/wordpress-develop/pull/5926/commits/96bd9926aa6e6e40f4dc770155c3093e5c0fe7ea)

**Review by swissspidy — APPROVED** [review]
Just from looking at the code, the changes appear sound to me.

Feels a bit ironic that we end up replacing constants with globals :)

File: src/wp-includes/load.php
Comment (mukeshpanchal27) [inline src/wp-includes/load.php]:
```suggestion
	if ( is_child_theme() ) {
```
Does that make sense? not blockers.

File: src/wp-includes/comment-template.php
Comment (mukeshpanchal27) [inline src/wp-includes/comment-template.php]:
```suggestion
```

Remove empty line.

**Review by mukeshpanchal27 — APPROVED** [review]
Thanks @joemcgill for the update. LGTM! Left some nit-pick.

**Comment by joemcgill** [comment]
@swissspidy 

> Feels a bit ironic that we end up replacing constants with globals :)

Agreed, and I feel the same. Essentially, we were using the old constants as a way to store the application state for template paths while the functions that are responsible for getting these values can change after the initial state is set. By keeping them as constants, any functions that relied on the old constants were not testable. However, trying to deprecate the constants in favor of using the functions directly caused several bugs due to the fact that these functions are used throughout the application in different ways than the state stored in the old constants and expect different values.

Perhaps there is a different pattern that would be preferable to using globals in this case, but I didn't want to over-engineer something in order to fix the bugs that have been reported.

**Comment by carlomanf** [comment]
> I didn't want to over-engineer something in order to fix the bugs that have been reported.

It looks over-engineered to me already, and I'd use stronger language than "ironic" to describe the replacement of constants with globals. I just don't see enough evidence that these "bugs" are actually bugs at all. What it looks like to me is the real bugs got coincidentally fixed and a small number of vocal theme/plugin authors complained because they were relying on the bugs and were forced to update their theme/plugin.

By the time this patch gets committed and deployed, which I hope it doesn't, anyone affected will have found other solutions already and it will serve no purpose other than to frustrate many other plugin/theme authors who want functions that behave in simple and consistent ways and are easy to learn. This kind of thing makes people want to stop authoring themes/plugins altogether, and most of them just leave quietly without making their frustrations known.

Some core decisions that I've seen recently have been so perplexing that it made me honestly question whether "shooing" plugin/theme authors is the intention. If so, then go right ahead and commit this, but I'm still perplexed. If not, maybe reflect on why it so often looks that way. It seems there is often too much attention on designing a solution and not enough effort to define the problem.

**Comment by metropoliscreative** [comment]
@joemcgill Thanks for your time on this PR - I think it works to solve the multisite issues we discussed in ticket [#60290](https://core.trac.wordpress.org/ticket/60290) and restores previously expected functionality. 

Will this make it into 6.5?

**Comment by carlomanf** [comment]
@metropoliscreative [I previously suggested an alternative for you in ticket 60290](https://core.trac.wordpress.org/ticket/60290#comment:3) that you could use immediately without a core patch. I would be interested to know whether it achieves what you're looking for.

**Comment by metropoliscreative** [comment]
> @metropoliscreative [I previously suggested an alternative for you in ticket 60290](https://core.trac.wordpress.org/ticket/60290#comment:3) that you could use immediately without a core patch. I would be interested to know whether it achieves what you're looking for.

@carlomanf Yes, thanks for your feedback. I understand you have strong opinions about this ticket and PR, but the goal here is to address unintended functionality changes from [[56635]](https://core.trac.wordpress.org/changeset/56635) related to `switch_to_blog()`. If you feel this long-standing functionality is in fact a bug, then please open another ticket to address it.

**Comment by carlomanf** [comment]
@metropoliscreative Let me quote the part I am referring to, in case you missed it:

> If I read your code correctly, it looks like you could easily solve the problem by appending user ID's that have posts to an array and calling `get_template_part` on each of them after you switch back to blog 2.

The point I am making is that 6.5 is more than a month away, a number of core committers above have expressed concern about the design of this patch, and there is potentially a way for you to independently solve the problem you encountered. Maybe you are already doing something like this, but it would assist everyone involved to find a way forward if you could confirm.

File: src/wp-includes/template.php
Comment (swissspidy) [reply src/wp-includes/template.php:684]: Adding `@access private` would be good.

At the moment a 6.4.x before 6.5.0 doesn't look likely.

Comment (joemcgill) [reply src/wp-includes/template.php:684]: For now, I've chosen not to mark this as private, given that it essentially replaces `wp_templating_constants()` which was also public.

File: src/wp-includes/load.php
Comment (joemcgill) [reply src/wp-includes/load.php]: Sure, makes sense to me 👍🏻

**Comment by joemcgill** [comment]
Merged in https://core.trac.wordpress.org/changeset/57685. Thanks all for the feedback and testing.

**Pull request closed without merging**
